### PR TITLE
Nagastone backward compatibility

### PR DESCRIPTION
**Pull request by Gordon-Frohman**
Put back old nagastone block
Separated new nagastone block into two new blocks - nagastone and nagastone head
Removed direction and facing fields from TileEntityTFNagastone. Replaced them with nagastone block metadata
Every old nagastone block will drop new nagastone itemblock when broken
Every old nagastone itemblock will place new nagastone when used
Removed old nagastone blocks from cretive menu and hid them from NEI

### CODE REVIEW

**Comment by Caedis** [comment]
Please remove any commented out code

File: src/main/java/twilightforest/block/BlockTFNagastone.java [hunk 1]
```diff
@@ -82,403 +55,317 @@ public int getRenderType() {
...
+            };
+        }
+        if (type == 12) {
+            // straight
+            // n/s
+            orient = switch (side) {
+                case 0, 1 -> 2; // u/d
+                case 2, 3 -> 1; // e/w
+                case 4, 5 -> 0;
+                default -> orient;
```
Comment (leagris) [inline src/main/java/twilightforest/block/BlockTFNagastone.java:187]: Please do not use magic numbers for directions. Use the ForgeDirection enum class. It also has helpers for rotations so no need to reinvent the wheel.

**Review by leagris — DISMISSED** [review]
Replace all numerical direction everywhere with `ForgeDirection` enum and replace all custom rotation logic by `ForgeDirection` helper methods.

**Comment by Gordon-Frohman** [comment]
> Replace all numerical direction everywhere with `ForgeDirection` enum and replace all custom rotation logic by `ForgeDirection` helper methods.

I can technically replace Facing with ForgeDirection, but not Direction: the enum should have an option for block facing sideways.
I'd rather keep my enums though: Direction and Facing are two different things, and using ForgeDirection for both of these would create unnecessary complications

File: src/main/java/twilightforest/block/BlockTFNagastone.java (hunk 1 above)
Comment (Gordon-Frohman) [reply src/main/java/twilightforest/block/BlockTFNagastone.java:187]: That's an original Benimatic's code, lol

**Comment by leagris** [comment]
> > Replace all numerical direction everywhere with `ForgeDirection` enum and replace all custom rotation logic by `ForgeDirection` helper methods.
> 
> I can technically replace Facing with ForgeDirection, but not Direction: the enum should have an option for block facing sideways. I'd rather keep my enums though: Direction and Facing are two different things, and using ForgeDirection for both of these would create unnecessary complications

With the very same principles as the ForgeDirection enum

Defining your own enum would prevent having to make sense of magic numbers, including for the meta data encoding. where the meta would be the ordinal value. For example: Nagastone heading UP and having its Yaw to North `NagastoneOrient.UP_NORTH`:
```java
// Direct use of the enum name in switch cases greatly improve readability while not affecting performance.
case NagastoneOrient.UP_NORTH:
```
And when encoding for the metadata value:
```java
{
    NagastoneOrient orient = NagastoneOrient.UP_NORTH; // Init value
    // ...

    // Get numerical value of orientation for block meta
    meta = orient.ordinal();
```

In your enum, you'd have an array of the different orientations to use the block meta as an index to get the matching enum entry.

**Comment by Gordon-Frohman** [comment]
> For example: Nagastone heading UP and having its Yaw to North `NagastoneOrient.UP_NORTH`:

That sounds like a really bad idea. I do not have even one check in code that would require both direction and yaw, so this enum would be basically useless. It almost always checks for direction, and then checks for yaw. Like here, for example:
```
switch (dir) {
                case DOWN:
                    iconYNeg = block.CROSS_SECTION;
                    iconYPos = block.CROSS_SECTION;
                    switch (face) {
                        default:
                        case EAST:
```
Using this kind of enum would require replacing `case DOWN:` with
```
case DOWN_NORTH:
case DOWN_SOUTH:
case DOWN_EAST:
case DOWN_WEST:
```

And "magic numbers" you are mentioning are not even in my code. That's old Benimatic's code, which is only left there for backward compatibility, has nothing to do with neither direction nor yaw and is controlled entirely by blocks' metadata. So both no need and reason to fix it

**Comment by leagris** [comment]
> > For example: Nagastone heading UP and having its Yaw to North `NagastoneOrient.UP_NORTH`:
> 
> That sounds like a really bad idea. I do not have even one check in code that would require both direction and yaw, so this enum would be basically useless. It almost always checks for direction, and then checks for yaw. Like here, for example:

Here is a part of your code:
```java
        if (type == 12) {
            // straight
            // n/s
            orient = switch (side) {
                case 0, 1 -> 2; // u/d
                case 2, 3 -> 1; // e/w
                case 4, 5 -> 0;
                default -> orient;
```

And now how you'd have it with enums:
```java
        if (type == STRAIGHT_NS) {
            orient = switch (side) {
                case UP,  DOWN -> UP_DOWN;
                case EAST, WEST -> EAST_WEST;
                case NORTH, SOUTH -> NORTH_SOUTH;
                default -> orient;
```

**Comment by Gordon-Frohman** [comment]
> Here is a part of your code:
> 
> ```java
>         if (type == 12) {
>             // straight
>             // n/s
>             orient = switch (side) {
>                 case 0, 1 -> 2; // u/d
>                 case 2, 3 -> 1; // e/w
>                 case 4, 5 -> 0;
>                 default -> orient;
> ```

This. Is. Not. My. Code.

Why do I even bother trying to explain something to you if you are not reading my messages?

**Comment by leagris** [comment]
> Why do I even bother trying to explain something to you if you are not reading my messages?

I honestly thought that the code you committed was yours: https://github.com/GTNewHorizons/twilightforest/blob/c57959a2ef278c36f08575edf7b2ef63b0c20e42/src/main/java/twilightforest/block/BlockTFNagastone.java#L180-L189

I think I must stop reviewing your code here.

**Comment by Gordon-Frohman** [comment]
> I honestly thought that the code you committed was yours

It is commited by me indeed, but that's actually a reupload of an old version of nagastone made by Benimatic (TF creator)
It is mentioned in the top message and I tried to explain it to you twice
It is left here only for backward compatibility, so there's no need to change it a lot, at least I don't think so
Sorry for being rude, I lost control of myself for a moment

**Review by leagris — APPROVED** [review]
Thank you @Gordon-Frohman 

All tests 100% OK
- runClent21: Placing body and head blocks all orientations and directions. Save/Reload all fine OK
- -runServer21: Same tests, disconnect/reconnect running server. Shutdown/Restart server. All ok

**Comment by Gordon-Frohman** [comment]
> Thank you @Gordon-Frohman

Thanks to you. And once again, I apologise for my behaviour

**Pull request merged**